pkg/agent/run.go

```diff
@@ -190,7 +204,44 @@ func Run(cmd *cobra.Command, args []string) {
 	}
 }
 
-func gatherAndOutputData(config CombinedConfig, preflightClient client.Client, dataGatherers map[string]datagatherer.DataGatherer) {
+// Creates an event recorder for the agent's Pod object. Expects the env var
+// POD_NAME to contain the pod name. Note that the RBAC rule allowing sending
+// events is attached to the pod's service account, not the impersonated service
+// account (venafi-connection).
+func newEventf(installNS string) (Eventf, error) {
+	restcfg, err := kubeconfig.LoadRESTConfig("")
+	if err != nil {
+		logs.Log.Fatalf("failed to load kubeconfig: %v", err)
+	}
+	scheme := runtime.NewScheme()
+	_ = corev1.AddToScheme(scheme)
+
+	var eventf Eventf
+	if os.Getenv("POD_NAME") == "" {
+		eventf = func(eventType, reason, msg string, args ...interface{}) {}
+		logs.Log.Printf("error messages will not show in the pod's events because the POD_NAME environment variable is empty")
+	} else {
+		podName := os.Getenv("POD_NAME")
+
+		eventClient, err := kubernetes.NewForConfig(restcfg)
+		if err != nil {
+			return eventf, fmt.Errorf("failed to create event client: %v", err)
+		}
+		broadcaster := record.NewBroadcaster()
+		broadcaster.StartRecordingToSink(&clientgocorev1.EventSinkImpl{Interface: eventClient.CoreV1().Events(installNS)})
+		eventRec := broadcaster.NewRecorder(scheme, corev1.EventSource{})
```

Comment:
I tried adding the component and node name to the event source and that made it clearer to the user what generated the event:

```diff
diff --git a/deploy/charts/venafi-kubernetes-agent/templates/deployment.yaml b/deploy/charts/venafi-kubernetes-agent/templates/deployment.yaml
index 0115c1a..e76573d 100644
--- a/deploy/charts/venafi-kubernetes-agent/templates/deployment.yaml
+++ b/deploy/charts/venafi-kubernetes-agent/templates/deployment.yaml
@@ -102,6 +102,10 @@ spec:
             valueFrom:
               fieldRef:
                 fieldPath: metadata.uid
+          - name: POD_NODE
+            valueFrom:
+              fieldRef:
+                fieldPath: spec.nodeName
           {{- if .Values.metrics.enabled }}
           ports:
             - containerPort: 8081
diff --git a/pkg/agent/run.go b/pkg/agent/run.go
index 0fa8038..55e7530 100644
--- a/pkg/agent/run.go
+++ b/pkg/agent/run.go
@@ -230,7 +230,7 @@ func newEventf(installNS string) (Eventf, error) {
                }
                broadcaster := record.NewBroadcaster()
                broadcaster.StartRecordingToSink(&clientgocorev1.EventSinkImpl{Interface: eventClient.CoreV1().Events(installNS)})
-               eventRec := broadcaster.NewRecorder(scheme, corev1.EventSource{})
+               eventRec := broadcaster.NewRecorder(scheme, corev1.EventSource{Component: "venafi-kuberentes-agent", Host: os.Getenv("POD_NODE")})
                eventf = func(eventType, reason, msg string, args ...interface{}) {
                        eventRec.Eventf(&corev1.Pod{ObjectMeta: v1.ObjectMeta{Name: podName, Namespace: installNS, UID: types.UID(os.Getenv("POD_UID"))}}, eventType, reason, msg, args...)
                }

```

```console
$ kubectl describe pod -n venafi -l app.kubernetes.io/name=venafi-kubernetes-agent
...
Events:
  Type     Reason      Age    From                     Message
  ----     ------      ----   ----                     -------
  Warning  PushingErr  2m10s  venafi-kubernetes-agent  retrying in 4m22.791991423s after error: post to server failed: while loading the VenafiConnection venafi/venafi-components: connection is not ready yet (building connection failed): chain element 1 (VCPOAuth) error: Post "https://api.venafi.eu.example/v1/oauth2/v2.0/9a0cab61-2b00-11ee-ba09-0733b0fe5adc/token": dial tcp: lookup api.venafi.eu.example on 34.118.224.10:53: no such host
```

```yaml
- apiVersion: v1
  count: 1
  eventTime: null
  firstTimestamp: "2024-10-11T16:19:00Z"
  involvedObject:
    apiVersion: v1
    kind: Pod
    name: venafi-kubernetes-agent-54f85c5d76-qvxrj
    namespace: venafi
    uid: 06cf3699-b9c5-4f0d-a790-e7e244d5148b
  kind: Event
  lastTimestamp: "2024-10-11T16:19:00Z"
  message: 'retrying in 1m6.931738453s after error: post to server failed: while loading
    the VenafiConnection venafi/venafi-components: connection is not ready yet (building
    connection failed): chain element 1 (VCPOAuth) error: Post "https://api.venafi.eu.example/v1/oauth2/v2.0/9a0cab61-2b00-11ee-ba09-0733b0fe5adc/token":
    dial tcp: lookup api.venafi.eu.example on 34.118.224.10:53: no such host'
  metadata:
    creationTimestamp: "2024-10-11T16:19:00Z"
    name: venafi-kubernetes-agent-54f85c5d76-qvxrj.17fd7254ce4b5454
    namespace: venafi
    resourceVersion: "3695"
    uid: dd371133-7680-45c9-aff1-3cf2baae9866
  reason: PushingErr
  reportingComponent: venafi-kuberentes-agent
  reportingInstance: gke-eu-cluster-1-default-pool-f9b6e891-htsp
  source:
    component: venafi-kuberentes-agent
    host: gke-eu-cluster-1-default-pool-f9b6e891-htsp
  type: Warning
kind: EventList
metadata: {}

```